### PR TITLE
Snapshot description field PR

### LINKED ISSUE #5580: Add `Notes` or equivalent field  ("Description"?) to `Snapshot` object
We want a field in snapshot metadata that allows a user to enter text that describes the snapshot. I'm not sure that `Notes` is the name we want for this field - `Description` is another name that comes to mind.

### LINKED ISSUE #5769: snapshot  create failure should not trigger a factory-reset

### LINKED ISSUE #5773: Verify snapshot name in `create` before starting wrapped-function

### PR DESCRIPTION
**Pull request by ericpromislow**
- Adds the description field to snapshots
- Simplifies the use of the function-wrapper for snapshot-create and -restore
- Splits `manager.Create` to `manager.ValidateName` and `manager.Create` so that
  names can be validated before we start manipulating the lock-file and backend-state

Fixes #5580 
Fixes #5769
Fixes #5773

### CODE REVIEW

File: src/go/rdctl/cmd/snapshot.go
Comment (jandubois) [inline src/go/rdctl/cmd/snapshot.go:78]: Why are we not restarting the backend here (only when the backend was running before we invoked the wrapped function)?

File: src/go/rdctl/cmd/snapshot.go
Comment (jandubois) [inline src/go/rdctl/cmd/snapshot.go:84]: I don't understand what this means.

The "keep in sync" explanation worries me: there should be only a single place that the state is kept (in the file), and the code in `rdctl` should simply notify the app that it needs to recheck the file because the status may have changed.

File: src/go/rdctl/cmd/snapshotList.go
Comment (jandubois) [inline src/go/rdctl/cmd/snapshotList.go:93]: I'm not sure if it wouldn't be better to just replace non-printing character in the description by spaces, trim leading and trailing spaces, and replace all runs of multiple spaces with a single space. And then treat all descriptions the same.

This would provide more information in the CLI output; the newlines are really only for the GUI code.

File: src/go/rdctl/cmd/snapshotList.go
Comment (jandubois) [inline src/go/rdctl/cmd/snapshotList.go:96]: And should we use a proper unicode ellipsis instead of appending 3 dots?

```suggestion
		if len(desc) > 61 {
			desc = desc[0:60] + "…"
```

The `60` feels somewhat arbitrary. At least when the output is a terminal, I would expect the output to use the full terminal width and not have a random cutoff point.

And I could make an argument that we should not truncate the output when it goes to a file (e.g. when you pipe into `less`) because the user can view the wrapped output, or scroll horizontally.

But we should open an enhancement request for this for 1.12.

**Review by jandubois — CHANGES_REQUESTED** [review]
First round of reviews

File: src/go/rdctl/cmd/snapshotList.go
Comment (jandubois) [reply src/go/rdctl/cmd/snapshotList.go:96]: > But we should open an enhancement request for this for 1.12.

I've created #5782 for this.

File: src/go/rdctl/cmd/snapshot.go
Comment (ericpromislow) [reply src/go/rdctl/cmd/snapshot.go:84]: This is an existing comment that got moved around, not part of this PR and not something I wrote.

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (ericpromislow) [reply bats/tests/snapshots/create-use-snapshot.bats:90]: It was stopped due to the error (which is fixed here) that trying to restore a nonexistent snapshot left the app in the STOPPED state. But I want it shutdown anyway for faster snapshot creation.

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (ericpromislow) [reply bats/tests/snapshots/create-use-snapshot.bats:92]: Part of this is in the test title: `@test "... and doesn't do factory-reset"`

Good point on rd-engine.json -- changed the assertion to `$PATH_CONFIG_FILE`.

And I do the existence test before running `snapshot create` because if I want to verify that
the failed `snapshot create` doesn't delete it. I have to do an indirect test like this.

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (ericpromislow) [reply bats/tests/snapshots/create-use-snapshot.bats:126]: I'm going with 60 characters for now. I'm not sure how straightforward it is in `go` to get the terminal width, or if stdout is bound to a terminal or tty. Agree with you that this is a separate issue -- you can always get the full description from the json view.

File: src/go/rdctl/cmd/snapshot.go
Comment (adamkpickering) [reply src/go/rdctl/cmd/snapshot.go:78]: > only when the backend was running before we invoked the wrapped function

Didn't we say that we wanted to try restarting the backend, assuming no error in the snapshot operation, whether we were able to shut it down? See #5674.

We need to look at this function in the context of the two places it is used:

Snapshot creation: when `manager.Create()` returns an error, the backend should be restarted and then the error should be returned.

Snapshot restoration: when `manager.Restore()` returns an error, it should do a factory reset and then return the error.

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (ericpromislow) [reply bats/tests/snapshots/create-use-snapshot.bats:138]: I don't see why it's necessary here - I control both the name and the description.

File: src/go/rdctl/cmd/snapshot.go
Comment (adamkpickering) [reply src/go/rdctl/cmd/snapshot.go:78]: Also, I still think it would be easier (both now and in the future when we make modifications) to do away with `wrapSnapshotOperation` and put the necessary code in `createSnapshot()` and `restoreSnapshot()`.

Comment (ericpromislow) [reply src/go/rdctl/cmd/snapshot.go:78]: This isn't an actual change in this PR -- you wouldn't see this in `git diff --color-moved`.

File: src/go/rdctl/cmd/snapshotList.go
Comment (ericpromislow) [reply src/go/rdctl/cmd/snapshotList.go:93]: Should be part of issue 5782 to deal with the cosmetic appearance of a description field in non-json output. Otherwise this PR will never get in.

File: src/go/rdctl/cmd/snapshot.go
Comment (adamkpickering) [reply src/go/rdctl/cmd/snapshot.go:84]: I agree that the way it works right now is not ideal. But this is the way it works right now (and it does work), and we are trying to release. Can we deal with this after 1.11?

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (jandubois) [reply bats/tests/snapshots/create-use-snapshot.bats]: Ok, then maybe remove the comment, if it is no longer accurate.

As I wrote, I agree that stopping the app is needed; just the comment looked weird.

File: src/go/rdctl/cmd/snapshot.go
Comment (adamkpickering) [inline src/go/rdctl/cmd/snapshot.go:88]: As I said in https://github.com/rancher-sandbox/rancher-desktop/pull/5715#discussion_r1364556070, I think we should keep the `config.GetConnectionInfo()` error handling code the same rather than adding `getConnectionInfo()` back.

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (jandubois) [reply bats/tests/snapshots/create-use-snapshot.bats:138]: It is not necessary right now, but it makes it more robust if the code gets refactored in the future.

Anyways, we can leave it for now.

File: src/go/rdctl/cmd/snapshotCreate.go
Comment (adamkpickering) [inline src/go/rdctl/cmd/snapshotCreate.go:55]: Why add all of this complexity when `tmutil addexclusion` is idempotent? This is a premature optimization, no?

File: src/go/rdctl/cmd/snapshotCreate.go
Comment (adamkpickering) [inline src/go/rdctl/cmd/snapshotCreate.go:61]: Error strings shouldn't end with newlines.

File: src/go/rdctl/cmd/snapshotCreate.go
Comment (adamkpickering) [inline src/go/rdctl/cmd/snapshotCreate.go:66]: Error strings shouldn't end with newlines.

File: bats/tests/snapshots/test-snapshot-list.bats
Comment (adamkpickering) [inline bats/tests/snapshots/test-snapshot-list.bats:31]: :+1: :+1:

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (adamkpickering) [inline bats/tests/snapshots/create-use-snapshot.bats:25]: My only real comment here is that I'm concerned that we have a tendency to grow existing bats tests rather than creating new ones. Shouldn't each .bats file contain a test for a relatively specific thing?

**Review by adamkpickering — CHANGES_REQUESTED** [review]

File: src/go/rdctl/pkg/snapshot/manager.go
Comment (jandubois) [reply src/go/rdctl/pkg/snapshot/manager.go:84]: A limit of 250 characters is implemented in https://github.com/rancher-sandbox/rancher-desktop/pull/5777/commits/d302ce426de697f89b4b913c2bda18a4e57182fc (part of #5777, not yet merged)

Comment (jandubois) [reply src/go/rdctl/pkg/snapshot/manager.go:84]: Sorry, that was a limit on the name length. The length of the description is unlimited. It is the user's problem if this becomes unwieldy in the UI. But you may want to auto-generate a description in a script from a manifest of all software and data included in the snapshot, and we should not artificially limit the size.

File: bats/tests/helpers/vm.bash
Comment (ericpromislow) [reply bats/tests/helpers/vm.bash:289]: No. Those three states are all stable states, where we know the backend is ready. We don't care about the state of kubernetes.

The other states mean that the backend is doing something else, but will eventually reach one of ```
       ERROR) return 0 ;;
        STARTED) return 0 ;;
        DISABLED) return 0 ;;
```

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (jandubois) [reply bats/tests/snapshots/create-use-snapshot.bats:25]: We will need to refactor them at some point.

However, we do want to combine closely related tests into a single file to reduce the number of times Rancher Desktop needs to be started and shutdown to keep running time as low as possible. It already takes 7 hours to run 4 different configurations of BATS on my old Mac Mini, and the time will only grow as we add tests.

But if tests are just aggregated, but not really related, and there is no shared startup savings, then we may want to split them into separate files.

File: src/go/rdctl/pkg/snapshot/manager_unix_test.go
Comment (ericpromislow) [reply src/go/rdctl/pkg/snapshot/manager_unix_test.go:77]: Good point, thx

File: src/go/rdctl/pkg/snapshot/manager.go
Comment (ericpromislow) [reply src/go/rdctl/pkg/snapshot/manager.go:65]: Whoa ... out-dated comment. I used to verify that `ValidateName(X)` was called before running `Create(X)`, but ran into problems with unit-testing.

Good point, I don't see why we would want to create more than one snapshot at a time either, so I removed the comment.

File: src/go/rdctl/cmd/snapshot.go
Comment (ericpromislow) [reply src/go/rdctl/cmd/snapshot.go:88]: The failing tests drove me to this solution. Refactoring this should be a separate issue.

File: src/go/rdctl/cmd/snapshot.go
Comment (ericpromislow) [reply src/go/rdctl/cmd/snapshot.go:27]: Now that the callback is just `func () error` I don't think there's a need for a name, which would show up only
in the defn of `wrapSnapshotOperation`

Not naming it is a little longer, but because it's used only once I don't see that we need to use a named type.

File: src/go/rdctl/cmd/snapshotCreate.go
Comment (ericpromislow) [reply src/go/rdctl/cmd/snapshotCreate.go:61]: yes

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (ericpromislow) [reply bats/tests/snapshots/create-use-snapshot.bats:25]: A lot of the tests I write build up some context, and then work with it.

I actually work on each test in an isolated file, but then plug it into the bigger bats test, if there is one, so it has the context it needs.

**Review by jandubois — APPROVED** [review]
Thanks, LGTM, but please wait for @adamkpickering to approve as well before merging.

File: src/go/rdctl/cmd/snapshot.go
Comment (adamkpickering) [reply src/go/rdctl/cmd/snapshot.go:86]: I created https://github.com/rancher-sandbox/rancher-desktop/issues/5792 to track this.

File: bats/tests/snapshots/create-use-snapshot.bats
Comment (adamkpickering) [reply bats/tests/snapshots/create-use-snapshot.bats:25]: > I actually work on each test in an isolated file, but then plug it into the bigger bats test, if there is one, so it has the context it needs.

If that's not an argument for smaller BATS tests, I don't know what is. We shouldn't have to move code around to develop it.

It's just hard to understand a 180 line BATS test.

But it's not relevant right now. Let's get this merged.

**Review by adamkpickering — APPROVED** [review]

**Pull request merged**
